mars/services/task/analyzer/fusion.py

```diff
@@ -12,101 +12,165 @@
+
+            if chunk.op in op_to_colors:
+                # colored
+                chunk_to_colors[chunk] = op_to_colors[chunk.op]
                 continue
-            selected = [v]
-            # add successors
-            while self._graph.count_predecessors(cur_node) == 1 and not isinstance(
-                cur_node.op, (VirtualOperand, Fetch)
-            ):
-                selected.append(cur_node)
-                if (
-                    self._graph.count_successors(cur_node) != 1
-                    or cur_node in result_chunk_set
-                ):
-                    break
+
+            predecessors = self.chunk_graph.predecessors(chunk)
+            pred_colors = {op_to_colors[pred.op] for pred in predecessors}
+            if len(predecessors) == 1 and chunk_to_is_broadcaster.get(predecessors[0]):
+                # predecessor is broadcaster, just allocate a new color
+                color = self._next_color()
+            elif len(pred_colors) == 1:
+                if self._can_color_same(chunk, predecessors):
+                    # predecessors have only 1 color, will color with same one
+                    color = next(iter(pred_colors))
                 else:
-                    cur_node = self._graph.successors(cur_node)[0]
-            if len(selected) > 1:
-                explored.update(selected)
-                fuses.append(list(selected))
-        return self._fuse_nodes(fuses)
+                    color = self._next_color()
+            else:
+                # has more than 1 color, color a new one
+                assert len(pred_colors) > 1
+                color = self._next_color()
+
+            op_to_colors[chunk.op] = chunk_to_colors[chunk] = color
+
+        # step 3: Propagate with reversed topological order,
+        # check a node with its inputs, if all inputs have different color with itself, skip;
+        # otherwise, if some of inputs have the same color, but some others have different color,
+        # color the input nodes with same one with a new color, and propagate to its inputs and so on.
+        for chunk in self.chunk_graph.topological_iter(reverse=True):
+            pred_colors = {
+                op_to_colors[pred.op]
+                for pred in self.chunk_graph.iter_successors(chunk)
+            }
+            chunk_color = chunk_to_colors[chunk]
+            if chunk_color in pred_colors and len(pred_colors) > 1:
+                # conflict
+                # color the successors with new colors
+                stack = []
+                for succ in self.chunk_graph.iter_successors(chunk):
+                    if chunk_to_colors[succ] == chunk_color:
+                        chunk_to_colors[succ] = op_to_colors[
+                            succ.op
+                        ] = self._next_color()
+                        stack.extend(self.chunk_graph.successors(succ))
+                # color the descendants with same color to the new one
```

Comment:
I'm thinking over this iteration-in-iteration implementation and get an idea without proof that one node may only be iterated at least twice. Maybe some comments needed here to show that the time complexity does not degrade.